TRUSTAUTH-28: Support for AJP Attributes

HP41 opened the pull request:
- Should allow use of grabbing auth data from AJP attributes which is much more secure than HTTP headers

HP41 committed:
Add Attributes Adapter

- Should allow use of grabbing auth data from AJP attributes which is much more secure than HTTP headers

tmortagne commented on xwiki-authenticator-trusted-adapters/xwiki-authenticator-trusted-attributes/src/main/java/org/xwiki/contrib/authentication/AttributesTrustedAuthenticationAdapter.java:
```diff
@@ -0,0 +1,197 @@
+ * Implementation of the {@link TrustedAuthenticationAdapter} for attributes authentication.
+ *
+ * @version $Id$
+ */
```
There should be a `@since 1.9.0`.

HP41 replied:
Added! 
PS: Headers Adapter does not have a version set

tmortagne replied:
>  PS: Headers Adapter does not have a version set

Yes, that's because it's been here from the start.

HP41 commented:
Please Note, I've not been able to test this. 

This is based on the same HTTPServeletRequest (XWikiRequest) which has getHeader and getAttribute.

This has been mostly a plain copy of the headers adapter but adjusted for `getAttribute`

tmortagne commented on xwiki-authenticator-trusted-adapters/xwiki-authenticator-trusted-attributes/src/main/java/org/xwiki/contrib/authentication/AttributesTrustedAuthenticationAdapter.java:
```diff
@@ -0,0 +1,193 @@
+            String encoding = configuration.getCustomProperty(CONFIG_ATTRIBUTE_ENCODING, null);
+            if (StringUtils.isNotBlank(encoding) && Charset.isSupported(encoding)) {
+                try {
+                    value = new String(value.getBytes("ISO-8859-1"), encoding);
```
It's not very clear to me why this is needed, it's supposed to be parsed already. At the very least, it deserves a comment to explain it.

HP41 replied:
Ah, am not sure if there needs to be a decoding akin to how Headers Adapter is done. What do you think? If it is not needed, please feel free to remove it

HP41 replied:
Happy to remove this section if it makes no sense. What do you think?

tmortagne replied:
As I said, I don't understand why it would be needed, but there must be a reason why you did this in the first place ? Was it just a copy/paste ?

On headers side, the reason seems to be https://jira.xwiki.org/browse/TRUSTAUTH-9. I would avoid this kind of hack (I'm honestly not really sure it's still accurate for headers) for attributes until proven that there is an actual need for it.

HP41 replied:
Perfect, that does simplify things. I've added a new commit.

Also added more safety around getAttribute casting (as it returns an Object and not String)

tmortagne commented:
> Please Note, I've not been able to test this.

I don't understand, it's not a need you have ?

tmortagne commented:
By the way, it would be nice to create an issue on https://jira.xwiki.org/browse/TRUSTAUTH for this new feature (we tend to use jira as changelog in XWiki contrib extensions).

HP41 commented:
> > Please Note, I've not been able to test this.
> 
> I don't understand, it's not a need you have ?

Yes, this is a nice to have need as AJP would align with best practices for Shibboleth.

HP41 commented:
> By the way, it would be nice to create an issue on https://jira.xwiki.org/browse/TRUSTAUTH for this new feature (we tend to use jira as changelog in XWiki contrib extensions).

I do not have a Jira account there but am working (as a client) with XWiki SAS team.

tmortagne commented:
> I do not have a Jira account there

Anyone can register an account there.

HP41 commented:
> > I do not have a Jira account there
> 
> Anyone can register an account there.

I've created Jira issues for both issues/PR now

HP41 committed:
Cleanup getAttribute

- Remove unnecessary encoding logic
- More safety logic around casting and trimming attribute

tmortagne commented on xwiki-authenticator-trusted-adapters/xwiki-authenticator-trusted-attributes/src/main/java/org/xwiki/contrib/authentication/AttributesTrustedAuthenticationAdapter.java:
```diff
@@ -0,0 +1,184 @@
+        } catch (ClassCastException e) {
+            logger.debug("Failed to cast attribute [{}] as string.", name, e);
+            return null;
+        } catch (NullPointerException e) {
```
Not a fan of this kind of (very slow) way to handle problems like these. Would be much cleaner to actively check if the value is a `String` before manipulating it. I also doubt you need to log the fact that an attribute is not set since it probably happen quite often and is something expected.

HP41 replied:
I updated it with a simpler flow!

HP41 commented:
Thank you very much!

tmortagne commented:
Thanks for your contribution! I just released the 1.9.0 version of the extension with it. Would be great if you could document this new feature on https://extensions.xwiki.org/xwiki/bin/view/Extension/Attributes%20Trusted%20Authentication%20Adapter/ (and https://extensions.xwiki.org/xwiki/bin/view/Extension/Trusted%20authentication%20framework if needed).

HP41 commented:
> Thanks for your contribution! I just released the 1.9.0 version of the extension with it. Would be great if you could document this new feature on https://extensions.xwiki.org/xwiki/bin/view/Extension/Attributes%20Trusted%20Authentication%20Adapter/ (and https://extensions.xwiki.org/xwiki/bin/view/Extension/Trusted%20authentication%20framework if needed).

Will have a look at it tomorrow.

In the meantime, I have had the chance to test this further and works fine so far!
